README


This archive contains the distribution Device-TMP102,
version 0.0.7:

  I2C interface to TMP102 temperature sensor

This software is Copyright (c) 2014 by Alex White.

This is free software, licensed under:

  The (three-clause) BSD License


This README file was generated by Dist::Zilla::Plugin::Readme v5.019.



Changes
==================================================
Changes from 2013-07-04 00:00:00 +0000 to present.
==================================================

------------------------------------------
version 0.0.7 at 2014-07-04 22:23:54 +0000
------------------------------------------

  Change: 12794d29956afa908c5c3d76c233f8b300ff0553
  Author: Alex White <VVu@geekfarm.org>
  Date : 2014-06-29 14:39:05 +0000

    handle error returned from I2C Device

    Also read temperature directly from I2CDeviceAddress. Apparently the
    rest of this code isn't really necessary and sometimes breaks things. 

------------------------------------------
version 0.0.6 at 2014-02-03 08:37:30 +0000
------------------------------------------

  Change: 9c2dffb6fe6260dbcdd1607e44c0cb868b1cfe72
  Author: Alex White <VVu@geekfarm.org>
  Date : 2014-02-03 00:37:30 +0000

    fix default perl path in shebang 

  Change: 4c4dc7c3d0d210b58a038074c4000c58254e352c
  Author: Alex White <VVu@geekfarm.org>
  Date : 2014-02-03 00:37:21 +0000

    fix some variable names 

  Change: e3184859f26678c15b04199d529b6fa07cc600b3
  Author: Alex White <VVu@geekfarm.org>
  Date : 2014-02-03 00:37:07 +0000

    set I2CBusDevicePath in test cases 

  Change: a9086889448d5e24213f6fe4d1b9db3e386b8d1f
  Author: Alex White <VVu@geekfarm.org>
  Date : 2014-02-02 23:27:52 +0000

    die if no temperature data received from sensor

    - created a separate method to set pointer register

    - added some more author test cases 

  Change: dd76a44b342f3bd54981ca78a8ae5887c203523a
  Author: Alex White <VVu@geekfarm.org>
  Date : 2014-02-02 23:26:21 +0000

    update pod docs, add SYNOPSIS 

  Change: 935f0cfc577154719910ba669499f0d9f785c325
  Author: Alex White <VVu@geekfarm.org>
  Date : 2014-02-02 23:19:15 +0000

    update copyright year to 2014 

  Change: 05bd6a1bc406a929aa40b2838cb201c7bfcbc0fb
  Author: Alex White <VVu@geekfarm.org>
  Date : 2014-02-02 23:18:21 +0000

    switch deprecated CriticTests to Test::Perl::Critic 

  Change: a4e40d525c3300c0ee5bdce2aad5a9e160bd2935
  Author: Alex White <VVu@geekfarm.org>
  Date : 2014-02-02 23:17:49 +0000

    add VERSION comment to bin/tmp102 

------------------------------------------
version 0.0.5 at 2013-12-10 07:23:43 +0000
------------------------------------------

  Change: 43af8a7464fd1f77bcaa872253d3df270738e40d
  Author: Alex White <VVu@geekfarm.org>
  Date : 2013-12-09 23:23:43 +0000

    bug fix to Device::TMP102

    Removing references to PCA9685, which was used as a model to create
    Device::TMP102. 

------------------------------------------
version 0.0.4 at 2013-12-10 06:16:09 +0000
------------------------------------------

  Change: 0fe71e2beac968fac31f8218dd7b9b8749ecd51d
  Author: Alex White <VVu@geekfarm.org>
  Date : 2013-12-09 22:16:09 +0000

    v0.0.4 

  Change: 05a2037e0eaad96539f851eb13b8f7782a832b59
  Author: Alex White <VVu@geekfarm.org>
  Date : 2013-12-09 21:56:50 +0000

    doc updates 

  Change: 557cc906305c5dbe42150930185788a0ec8b4fda
  Author: Alex White <VVu@geekfarm.org>
  Date : 2013-12-09 20:38:28 +0000

    docs link to Device::Temperature::TMP102 

------------------------------------------
version 0.0.3 at 2013-12-10 01:58:46 +0000
------------------------------------------

  Change: 9bd6c81cfe93be5c35eb345bd490a9fa9e449330
  Author: Alex White <VVu@geekfarm.org>
  Date : 2013-12-09 17:58:46 +0000

    install tmp102 script

    - renamed script

    - moved to 'bin/' directory

    - add ExecDir to dist.ini 

  Change: e4cb0c2bae77e0a5b932cfae0b468da525d9b1fb
  Author: Alex White <VVu@geekfarm.org>
  Date : 2013-12-08 21:29:37 +0000

    update links in docs

    - sparkfun item 11931 replaces 9418

    - add link to github page 

------------------------------------------
version 0.0.2 at 2013-12-09 05:16:01 +0000
------------------------------------------

  Change: 87ec1cb7a30cab8c148fea32fcf26ddf45eb5918
  Author: Alex White <VVu@geekfarm.org>
  Date : 2013-12-08 21:16:01 +0000

    v0.0.2 

  Change: 03fe8b7d6026d2254f8940f9986dbb9355c7c709
  Author: Alex White <VVu@geekfarm.org>
  Date : 2013-12-08 21:14:56 +0000

    switch to Dist::Zilla::Plugin::ChangelogFromGit 

  Change: 25fe1af8ee120f1c97b97628a52c3313f3a043a3
  Author: Alex White <VVu@geekfarm.org>
  Date : 2013-12-08 21:10:08 +0000

    remove duplicate link 

  Change: 4ee28f5d832bba084e5959cc6ab428e3a7d9288b
  Author: Alex White <VVu@geekfarm.org>
  Date : 2013-12-08 21:09:49 +0000

    eliminate temporary variable 

  Change: 5dd4380586b91b73547ada8a34bdbd32ed569c57
  Author: Alex White <VVu@geekfarm.org>
  Date : 2013-12-08 20:41:55 +0000

    update from sample TMP102 test code 

  Change: 55a622eefc3011a061a35c8a63a0e956690c7b5a
  Author: Alex White <VVu@geekfarm.org>
  Date : 2013-12-08 20:40:53 +0000

    update pod docs 

------------------------------------------
version 0.0.1 at 2013-12-09 04:10:24 +0000
------------------------------------------

  Change: 38961246e4b03cf729d0dd65daeea3800340696a
  Author: Alex White <VVu@geekfarm.org>
  Date : 2013-12-08 20:10:24 +0000

    fix another pod typo 

  Change: fccb3c80b231994d105ee446365cef03e85072f1
  Author: Alex White <VVu@geekfarm.org>
  Date : 2013-12-08 20:09:11 +0000

    add stopwords for pod spelling test 

  Change: 2413f5f81318442b153b3a5f7870a2ef216643c5
  Author: Alex White <VVu@geekfarm.org>
  Date : 2013-12-08 20:08:45 +0000

    fix typo in pod 

  Change: f1bf54392408ef88bcd13fb1e9e9eef84fdd1094
  Author: Alex White <VVu@geekfarm.org>
  Date : 2013-12-08 20:02:07 +0000

    adding Changes file for initial release 

  Change: 89da742e38d9670c118aab9a269ab204252b81fc
  Author: Alex White <VVu@geekfarm.org>
  Date : 2013-12-08 20:01:37 +0000

    add .gitignore 

  Change: 501e35f23b17ce911c5ea47c49c6fb3d48e15a47
  Author: Alex White <VVu@geekfarm.org>
  Date : 2013-12-08 19:59:00 +0000

    make command line script executable 

  Change: 1369d96a0824b8bc4b66fe4a8b1af3c0284b24ec

Makefile.PL

# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v5.019.
use strict;
use warnings;



use ExtUtils::MakeMaker 6.30;



my %WriteMakefileArgs = (
  "ABSTRACT" => "I2C interface to TMP102 temperature sensor",
  "AUTHOR" => "Alex White <VVu\@geekfarm.org>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.30"
  },
  "DISTNAME" => "Device-TMP102",
  "EXE_FILES" => [
    "bin/tmp102"
  ],
  "LICENSE" => "bsd",
  "NAME" => "Device::TMP102",
  "PREREQ_PM" => {
    "Device::SMBus" => 0,
    "Moose" => 0,
    "POSIX" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "Test::More" => 0
  },
  "VERSION" => "0.0.7",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "Device::SMBus" => 0,
  "Moose" => 0,
  "POSIX" => 0,
  "Test::More" => 0,
  "strict" => 0,
  "warnings" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);





lib/Device/Temperature/TMP102.pm
package Device::Temperature::TMP102;
use Moose;

our $VERSION = '0.0.7'; # VERSION

extends 'Device::SMBus';

has '+I2CDeviceAddress' => (
    is      => 'ro',
    default => 0x48,
);

has debug => (
    is      => 'ro',
    default => 0,
);

sub getTemp {
    my ( $self ) = @_;

    my $results = $self->readWordData( $self->I2CDeviceAddress );

    unless ( $results ) {
        die( "ERROR: failed to get temperature reading" );
    }

    if ( $results eq "-1" ) {
        die( "ERROR: got back an error code from I2C Device" );
    }

    return $self->convertTemp( $results );
}

sub convertTemp {
    my ( $self, $value ) = @_;

    my $lsb = ( $value & 0xff00 );
    $lsb = $lsb >> 8;

    my $msb = $value & 0x00ff;

    printf( "results: %04x\n", $value ) if $self->debug;
    printf( "msb:     %02x\n", $msb )   if $self->debug;
    printf( "lsb:     %02x\n", $lsb )   if $self->debug;

    my $temp = ( $msb << 8 ) | $lsb;

    # The TMP102 temperature registers are left justified, correctly
    # right justify them
    $temp = $temp >> 4;

    # test for negative numbers
    if ( $temp & ( 1 << 11 ) ) {

        # twos compliment plus one, per the docs
        $temp = ~$temp + 1;

        # keep only our 12 bits
        $temp &= 0xfff;

        # negative
        $temp *= -1;
    }

    # convert to a celsius temp value
    $temp = $temp / 16;

    return $temp;
}

__PACKAGE__->meta->make_immutable;

1;

__END__

=pod

=head1 NAME

Device::Temperature::TMP102 - I2C interface to TMP102 temperature sensor using Device::SMBus

=head1 SYNOPSIS

  use Device::Temperature::TMP102;

  my $device = shift @ARGV || '/dev/i2c-1';

  my $dev = Device::Temperature::TMP102->new( I2CBusDevicePath => $device );

  my $temp = $dev->getTemp();

  print "Temp:\n";
  printf ( "\t%2.2f C\n", $temp );
  printf ( "\t%2.2f F\n", $temp * 1.8 + 32 );


=head1 DESCRIPTION

Read temperature for a TMP102 temperature sensor over I2C.

This library correctly handles temperatures below freezing (0 degrees Celsius).

=head1 TROUBLESHOOTING

Check for your device on i2cbus 1 using the i2cdetect command, e.g.:

  $ i2cdetect -y 1

       0  1  2  3  4  5  6  7  8  9  a  b  c  d  e  f
  00:          -- -- -- -- -- -- -- -- -- -- -- -- --
  10: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- --
  20: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- --
  30: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- --
  40: -- -- -- -- -- -- -- -- 48 -- -- -- -- -- -- --
  50: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- --
  60: -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- --
  70: -- -- -- -- -- -- -- --

This indicates my TMP102 chip address is 0x48 on i2cbus 1.

Use the command-line tool i2cget to try to get a 12-bit reading from
the device, e.g.:

  $ i2cget -y 1 0x48 0x00 w
  0x2003

Refer to the documentation on L<Device::SMBus> for information on
enabling the i2c driver.

I also found this page to be helpful:

  http://donalmorrissey.blogspot.co.uk/2012/09/raspberry-pi-i2c-tutorial.html

In the process of testing this on raspberry pi, I saw this error:

  perl: symbol lookup error: .../Device/SMBus/SMBus.so: undefined symbol: i2c_smbus_write_byte

The fix was to install the package libi2c-dev.

=head1 SEE ALSO

  https://www.sparkfun.com/products/11931

  https://www.sparkfun.com/datasheets/Sensors/Temperature/tmp102.pdf

  http://donalmorrissey.blogspot.com/2012/09/raspberry-pi-i2c-tutorial.html

  https://github.com/sparkfun/Digital_Temperature_Sensor_Breakout_-_TMP102

=head1 SOURCE

With code and comments taken from example code for the ATmega328:

  http://www.sparkfun.com/datasheets/Sensors/Temperature/tmp102.zip

  /*
    TMP Test Code
    5-31-10
    Copyright Spark Fun Electronics© 2010
    Nathan Seidle

    Example code for the TMP102 11-bit I2C temperature sensor

    You will need to connect the ADR0 pin to one of four places. This
    code assumes ADR0 is tied to VCC.  This results in an I2C address
    of 0x93 for read, and 0x92 for write.

    This code assumes regular 12 bit readings. If you want the
    extended mode for higher temperatures, the code will have to be
    modified slightly.

  */

=head1 VERSION

=head1 ATTRIBUTES

=head2 I2CDeviceAddress

Contains the I2CDevice Address for the bus on which your TMP102 is
connected. The default value is 0x48.

=head1 METHODS

=head2 getTemp()

    $self->getTemp()

Returns the current temperature, in degrees Celsius.

=head2 convertTemp()

    $self->convertTemp( $reading )

Given a value read from the TMP102, convert the value to degrees
Celsius.

=head1 LICENSE

This software is Copyright (c) 2014 by Alex White.

This is free software, licensed under:

  The (three-clause) BSD License

The BSD License

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are
met:

  * Redistributions of source code must retain the above copyright
    notice, this list of conditions and the following disclaimer.

  * Redistributions in binary form must reproduce the above copyright
    notice, this list of conditions and the following disclaimer in the
    documentation and/or other materials provided with the distribution.

  * Neither the name of Alex White nor the names of its
    contributors may be used to endorse or promote products derived from
    this software without specific prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR
CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
